README
phonenumber
===========

An R package that converts English letters to numbers or numbers to English 
letters as on a telephone keypad. When converting letters to numbers, a 
character vector is returned with "A," "B," or "C" becoming 2, "D," "E", or "F" becoming 3, etc. When converting numbers to letters, a character vector is 
returned with multiple elements (i.e., "2" becomes a vector of "A," "B," and \
"C").

NEWS.md
# phonenumber 0.2.3 - DN38416

- add rmarkdown as a dependency for vignette
- update project and other URLs

# phonenumber 0.2.2 - machine phase matter

- tweaks to vignette and documentation

# phonenumber 0.2.1 - phased plasma rifle in the 40-watt range

- change for loops to lapply
- remove ineffective error-checking

# phonenumber 0.2.0 - memetic polyalloy

- add user determination of whether to map "Q" and "Z" to 7 and 9 or to 0
- add additional tests due to new functionality
- add comments to existing examples
- add additional examples due to new functionality
- add package help file
- fix typos and add title to vignette
- add URLs to DESCRIPTION
- updates to README

# phonenumber 0.1.0 - hyperalloy combat chassis

- first version of phonenumber


tests/testthat.R
library(testthat)
library(phonenumber)

test_check("phonenumber")


tests/testthat/test-letterToNumber.R
library(phonenumber)
context("letterToNumber")

test_that("letterToNumber returns the correct numbers", {
    expect_equal(letterToNumber("a"), "2")
    expect_equal(letterToNumber("b"), "2")
    expect_equal(letterToNumber("c"), "2")
    expect_equal(letterToNumber("d"), "3")
    expect_equal(letterToNumber("e"), "3")
    expect_equal(letterToNumber("f"), "3")
    expect_equal(letterToNumber("g"), "4")
    expect_equal(letterToNumber("h"), "4")
    expect_equal(letterToNumber("i"), "4")
    expect_equal(letterToNumber("j"), "5")
    expect_equal(letterToNumber("k"), "5")
    expect_equal(letterToNumber("l"), "5")
    expect_equal(letterToNumber("m"), "6")
    expect_equal(letterToNumber("n"), "6")
    expect_equal(letterToNumber("o"), "6")
    expect_equal(letterToNumber("p"), "7")
    expect_equal(letterToNumber("q"), "7")
    expect_equal(letterToNumber("q", qz = 0), "0")
    expect_equal(letterToNumber("r"), "7")
    expect_equal(letterToNumber("s"), "7")
    expect_equal(letterToNumber("t"), "8")
    expect_equal(letterToNumber("u"), "8")
    expect_equal(letterToNumber("v"), "8")
    expect_equal(letterToNumber("w"), "9")
    expect_equal(letterToNumber("x"), "9")
    expect_equal(letterToNumber("y"), "9")
    expect_equal(letterToNumber("z"), "9")
    expect_equal(letterToNumber("z", qz = 0), "0")
    expect_equal(letterToNumber("A"), "2")
    expect_equal(letterToNumber("B"), "2")
    expect_equal(letterToNumber("C"), "2")
    expect_equal(letterToNumber("D"), "3")
    expect_equal(letterToNumber("E"), "3")
    expect_equal(letterToNumber("F"), "3")
    expect_equal(letterToNumber("G"), "4")
    expect_equal(letterToNumber("H"), "4")
    expect_equal(letterToNumber("I"), "4")
    expect_equal(letterToNumber("J"), "5")
    expect_equal(letterToNumber("K"), "5")
    expect_equal(letterToNumber("L"), "5")
    expect_equal(letterToNumber("M"), "6")
    expect_equal(letterToNumber("N"), "6")
    expect_equal(letterToNumber("O"), "6")
    expect_equal(letterToNumber("P"), "7")
    expect_equal(letterToNumber("Q"), "7")
    expect_equal(letterToNumber("Q", qz = 0), "0")
    expect_equal(letterToNumber("R"), "7")
    expect_equal(letterToNumber("S"), "7")
    expect_equal(letterToNumber("T"), "8")
    expect_equal(letterToNumber("U"), "8")
    expect_equal(letterToNumber("V"), "8")
    expect_equal(letterToNumber("W"), "9")
    expect_equal(letterToNumber("X"), "9")
    expect_equal(letterToNumber("Y"), "9")
    expect_equal(letterToNumber("Z"), "9")
    expect_equal(letterToNumber("Z", qz = 0), "0")
    expect_equal(letterToNumber("1"), "1")
    expect_equal(letterToNumber("0"), "0")    
    expect_equal(letterToNumber("-"), "-")
    expect_equal(letterToNumber("#"), "-")
    expect_equal(letterToNumber("!"), "-")
    expect_equal(letterToNumber(" "), "-")
    expect_equal(letterToNumber("adZ 123"), "239-123")    
    expect_equal(letterToNumber("adZ 123", qz = 0), "230-123")    
})

tests/testthat/test-numberToLetter.R
library(phonenumber)
context("numberToLetter")

test_that("numberToLetter returns the correct letters", {
    expect_equal(numberToLetter("2"), c("A", "B", "C"))
    expect_equal(numberToLetter("3"), c("D", "E", "F"))
    expect_equal(numberToLetter("4"), c("G", "H", "I"))
    expect_equal(numberToLetter("5"), c("J", "K", "L"))
    expect_equal(numberToLetter("6"), c("M", "N", "O"))
    expect_equal(numberToLetter("7"), c("P", "Q", "R", "S"))
    expect_equal(numberToLetter("7", qz = 0), c("P", "R", "S"))
    expect_equal(numberToLetter("8"), c("T", "U", "V"))
    expect_equal(numberToLetter("9"), c("W", "X", "Y", "Z"))
    expect_equal(numberToLetter("9", qz = 0), c("W", "X", "Y"))
    expect_equal(numberToLetter("1"), "1")
    expect_equal(numberToLetter("0"), "0")
    expect_equal(numberToLetter("0", qz = 0), c("Q", "Z"))
    expect_equal(numberToLetter("-"), "-")
    expect_equal(numberToLetter("#"), "-")
    expect_equal(numberToLetter("!"), "-")
    expect_equal(numberToLetter(" "), "-")
    expect_equal(numberToLetter("911!"), c("W11-", "X11-", "Y11-", "Z11-"))    
    expect_equal(numberToLetter("911!", qz=0), c("W11-", "X11-", "Y11-"))    
    expect_equal(numberToLetter("000"), "000")
    expect_equal(numberToLetter("000", qz = 0), c("QQQ", "QQZ", "QZQ", "QZZ", 
                                                  "ZQQ", "ZQZ", "ZZQ", "ZZZ"))
})

R/letterToNumber.R
#' 
#' Convert letters to numbers as on a telephone's keypad
#' 
#' Take a character vector and convert it to the equivalent number sequence
#' from a telephone's key pad
#' 
#' @param value An input value as a character vector with one element (a string)
#' @param qz Whether to assign q and z to zero (qz = 0) or not (any other value)
#' 
#' @export
#' 
#' @examples
#' # Convert an alphabetic string can be converted directly (with 
#' # non-alphanumeric characters replaced by dashes)
#' 
#' letterToNumber("R functions") # returns "7-386284667"
#' 
#' # Of course, vectors containing strings can also be converted
#' 
#' string <- "Phone Number"
#' letterToNumber(string) # returns "74663-686237"
#' 
#' # Alphanumeric strings can also be converted with numbers being returned as
#' # is
#' 
#' letterToNumber("Jenny's number is 867-5309") # returns "53669-7-686237-47-867-5309"
#' 
#' # Specifying qz = 0 maps "q" and "z" to 0 instead of 7 and 9
#' 
#' letterToNumber("qz") # returns "79"
#' letterToNumber("qz", qz = 0) # returns ("00")
#' 
#' @return A character vector of numbers and dashes based on value

letterToNumber <- function(value, qz = 1) {
    value <- as.character(value)
    value <- gsub("[^A-Za-z0-9]", "-", value)
    value <- toupper(value)
    valueSplit <- strsplit(value, "")[[1]]
    numString <- as.character()
    mapphone <- function(char) { 
        if (qz == 0 && (char == LETTERS[17] | char == LETTERS[26])) { "0" } 
        else { ifelse(is.element(char, LETTERS[1:3]), "2",
               ifelse(is.element(char, LETTERS[4:6]), "3",
               ifelse(is.element(char, LETTERS[7:9]), "4",
               ifelse(is.element(char, LETTERS[10:12]), "5",
               ifelse(is.element(char, LETTERS[13:15]), "6",
               ifelse(is.element(char, LETTERS[16:19]), "7",
               ifelse(is.element(char, LETTERS[20:22]), "8",
               ifelse(is.element(char, "-") | 
                      suppressWarnings(!is.na(as.numeric(char))), char, "9")
               ))))))) 
        } 
        } 
    numString <- lapply(valueSplit, mapphone)
    return(paste0(numString, collapse = ""))
}

R/numberToLetter.R
#' 
#' Convert numbers to letters as on a telephone's keypad
#' 
#' Take a character vector (i.e., a telephone number) and convert it to all
#' all possible letter combinations as on from a telephone's key pad
#' 
#' @param value An input value as a character vector with one element (a string)
#' @param decreasing Whether to sort the results in alphabetical order or not
#' @param qz Whether to assign q and z to zero (qz = 0) or not (any other value)
#' 
#' @export
#' 
#' @examples
#' # Convert a string or a vector of numeric characters
#'
#' numberToLetter("911") # returns "W11" "X11" "Y11" "Z11"
#' x <- "911"
#' numberToLetter(x) # also returns "W11" "X11" "Y11" "Z11"
#' 
#' # Convert a number directly
#' 
#' numberToLetter(911) # also returns "W11" "X11" "Y11" "Z11"
#' 
#' # Convert an alphanumeric string (letters are returned as is and
#' # non-alphanumeric characters are returned as dashes)
#' 
#' numberToLetter("dial 911!") # returns "DIAL-W11-" "DIAL-X11-" "DIAL-Y11-" "DIAL-Z11-"
#'
#' # Specifying qz = 0 maps "q" and "z" to 0 instead of 7 and 9
#' 
#' numberToLetter("000") # returns "000" 
#' numberToLetter("000", qz = 0) # returns "QQQ" "QQZ" "QZQ" "QZZ" "ZQQ" "ZQZ" "ZZQ" "ZZZ"
#' 
#' @return A character vector of letters and dashes based on value

numberToLetter <- function(value, decreasing = FALSE, qz = 1) {
    value <- as.character(value)
    value <- gsub("[^A-Za-z0-9]", "-", value)
    value <- toupper(value)
    valueSplit <- strsplit(value, "")[[1]]    
    strList <- list()
    phonemap <- function(char) { if (char == "2") { strsplit(LETTERS[1:3], "") 
        } else { if (char == "3") { strsplit(LETTERS[4:6], "") 
        } else { if (char == "4") { strsplit(LETTERS[7:9], "") 
        } else { if (char == "5") { strsplit(LETTERS[10:12], "") 
        } else { if (char == "6") { strsplit(LETTERS[13:15], "") 
        } else { if (char == "7") { strsplit(LETTERS[16:19], "") 
        } else { if (char == "8") { strsplit(LETTERS[20:22], "") 
        } else { if (char == "9") { strsplit(LETTERS[23:26], "") } else { char } 
        }}}}}}}
        }
    letterSplit <- function(char) { if (qz == 0) {if (char == "0") { 
        list(LETTERS[17], LETTERS[26]) } else { if (char == "7") { 
            c(LETTERS[16], strsplit(LETTERS[18:19], "")) } else { 
                if (char == "9") { strsplit(LETTERS[23:25], "") 
                } else { phonemap(char) }}}} else { phonemap(char) } 
        }
    strList <- lapply(valueSplit, letterSplit)
    strDF <- expand.grid(strList)
    lettString <- do.call(paste0, strDF[1:ncol(strDF)])
    lettString <- sort(lettString, decreasing)
    return(lettString)
}